### PR TITLE
Add `j.u.c.Flow`

### PR DESCRIPTION
**Pull request by armanbilge**
https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.html

Since they are just interfaces, what are we looking for in terms of tests?

### CODE REVIEW

**Comment by sjrd** [comment]
The tests we do for pure interfaces are: create instances of it, "upcast" them to the interface type, call methods. This is to make sure that the binary API of the methods declared in the interfaces are actually.correct.

See for example https://github.com/scala-js/scala-js/blob/v1.12.0/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/function/LongSupplierTest.scala

File: javalib/src/main/scala/java/util/concurrent/Flow.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * additional information regarding copyright ownership.
+ */
+
+package java.util.concurrent
+
+object Flow {
+
+  trait Processor[T, R] extends Subscriber[T] with Publisher[R]
+
+  trait Publisher[T] {
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/concurrent/Flow.scala:19]: We do put `@FunctionalInterface` when the Java interface has it.

File: javalib/src/main/scala/java/util/concurrent/Flow.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    def subscribe(subscriber: Subscriber[T]): Unit
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/concurrent/Flow.scala:20]:
```suggestion
    def subscribe(subscriber: Subscriber[_ >: T]): Unit
```

File: javalib/src/main/scala/java/util/concurrent/Flow.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  trait Subscriber[T] {
+    def onComplete(): Unit
+    def onError(throwable: Throwable): Unit
+    def onNext(item: T): Unit
+    def onSubscribe(subscription: Subscription): Unit
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/concurrent/Flow.scala:27]: Please order the methods according the "Method Detail" section of the JavaDoc (not by alphabetical order).

File: javalib/src/main/scala/java/util/concurrent/Flow.scala [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  trait Subscription {
+    def cancel(): Unit
+    def request(n: Long): Unit
+  }
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/concurrent/Flow.scala:33]: Same here (so swap them in this case).

**Review by sjrd — CHANGES_REQUESTED** [review]

File: test-suite/shared/src/test/require-jdk11/org/scalajs/testsuite/javalib/util/concurrent/FlowTest.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    val subscription = makeSubscription()
+    subscription.request(42)
+    subscription.cancel()
+  }
+
+}
+
+object FlowTest {
+
+  def makeProcessor[T, R](): Flow.Processor[T, R] =
```
Comment (sjrd) [inline test-suite/shared/src/test/require-jdk11/org/scalajs/testsuite/javalib/util/concurrent/FlowTest.scala:55]:
```suggestion
  def makeProcessor[T, R](): Flow.Processor[T, R] = {
```
Same for all the `makeX` methods.

According to our coding style:
https://github.com/scala-js/scala-js/blob/main/CODINGSTYLE.md#braces

**Review by sjrd — CHANGES_REQUESTED** [review]
Looks good except for some missing braces in the tests, to adhere to our coding style.

**Comment by armanbilge** [comment]
Oh, I forgot to add `defaultBufferSize()` 😅 

https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.html#defaultBufferSize--

**Review by sjrd — APPROVED** [review]

**Comment by armanbilge** [comment]
Update: thanks to this interface is now trivial for users to consume FS2 streams in Laminar/Airstream, without either library knowing about each other. I think that's really cool :)

- https://github.com/typelevel/fs2/pull/3160
- https://github.com/raquo/Airstream/pull/122

**Pull request merged**